jdisc_http_service/src/main/java/com/yahoo/jdisc/http/server/jetty/HttpResponseStatisticsCollector.java

```diff
@@ -0,0 +1,199 @@
+import org.eclipse.jetty.server.handler.HandlerWrapper;
+import org.eclipse.jetty.util.FutureCallback;
+import org.eclipse.jetty.util.component.Graceful;
+
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * HttpResponseStatisticsCollector collects statistics about HTTP response types aggregated by category (1xx, 2xx, etc). It is similar to
+ * {@link org.eclipse.jetty.server.handler.StatisticsHandler} with the distinction that this class collects response type statistics grouped
+ * by HTTP method and only collects the numbers that are reported as metrics from Vespa.
+ *
+ * @author ollivir
+ */
+public class HttpResponseStatisticsCollector extends HandlerWrapper implements Graceful {
+    private final AtomicReference<FutureCallback> shutdown = new AtomicReference<>();
+
+    public static enum HttpMethod {
+        GET, PATCH, POST, PUT, REMOVE, UNKNOWN
+    }
+
+    private static final String[] HTTP_RESPONSE_GROUPS = { Metrics.RESPONSES_1XX, Metrics.RESPONSES_2XX, Metrics.RESPONSES_3XX,
+            Metrics.RESPONSES_4XX, Metrics.RESPONSES_5XX };
+
+    private final AtomicLong inFlight = new AtomicLong();
+    private final LongAdder statistics[][];
+
+    public HttpResponseStatisticsCollector() {
+        super();
+        statistics = new LongAdder[HttpMethod.values().length][];
+        for (int method = 0; method < statistics.length; method++) {
+            statistics[method] = new LongAdder[HTTP_RESPONSE_GROUPS.length];
+            for (int group = 0; group < HTTP_RESPONSE_GROUPS.length; group++) {
+                statistics[method][group] = new LongAdder();
+            }
+        }
+    }
+
+    private final AsyncListener completionWatcher = new AsyncListener() {
+        @Override
+        public void onTimeout(AsyncEvent event) throws IOException {
+        }
+
+        @Override
+        public void onStartAsync(AsyncEvent event) throws IOException {
+            event.getAsyncContext().addListener(this);
+        }
+
+        @Override
+        public void onError(AsyncEvent event) throws IOException {
+        }
+
+        @Override
+        public void onComplete(AsyncEvent event) throws IOException {
+            HttpChannelState state = ((AsyncContextEvent) event).getHttpChannelState();
+            Request request = state.getBaseRequest();
+
+            observeEndOfRequest(request, null);
+        }
+    };
+
+    @Override
+    public void handle(String path, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
```

Comment:
Consider adding a note that most the logic is taken from Jetty's `StatisticsHandler`.

Reply:
Will be added in next commit.